build-tools/packages/build-cli/src/commands/generate/sourceEntrypoints.ts

```diff
@@ -0,0 +1,386 @@
+}
+
+/**
+ * Read and parse tsconfig under `outDir`
+ * @returns rootDir
+ */
+async function readOutDirTsConfig(outDir: string): Promise<string> {
+	const formatOutDir = formatPath(outDir);
+
+	// Read tsconfig under `${outDir}`
+	const tsConfigContent = await fs.readFile(`${formatOutDir}tsconfig.json`, {
+		encoding: "utf8",
+	});
+
+	if (tsConfigContent === undefined) {
+		throw new Error(`tsconfig.json not found in ${formatOutDir}`);
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+	const tsconfig = JSON5.parse(tsConfigContent);
+
+	let rootDir = "";
+
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+	if (tsconfig.compilerOptions?.rootDir !== undefined) {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
+		rootDir = tsconfig.compilerOptions?.rootDir;
+	}
+
+	if (rootDir.length === 0 || rootDir === undefined) {
+		throw new Error(`No rootDir defined in ${formatOutDir}tsconfig.json`);
+	}
+
+	return formatPath(rootDir);
+}
+
+/**
+ * Returns a map of source paths to `ApiTag` levels.
+ */
+async function mapSourcePathToApiTag(outDir: string): Promise<Map<string, ApiTag>> {
+	const rootDir = await readOutDirTsConfig(outDir);
+
+	const outFileSuffix = ".ts";
+
+	return new Map([
+		[`${rootDir}${ApiLevel.alpha}${outFileSuffix}`, ApiTag.alpha],
+		[`${rootDir}${ApiLevel.beta}${outFileSuffix}`, ApiTag.beta],
+		[`${rootDir}${ApiLevel.public}${outFileSuffix}`, ApiTag.public],
+		[`${rootDir}${ApiLevel.legacy}${outFileSuffix}`, ApiTag.legacy],
+	]);
+}
+
+/**
+ * Resolves a mapping of `ApiTag` levels to their modified export paths.
+ */
+function getOutputConfiguration(
+	mapApiTagToSourcePath: ReadonlyMap<string, ApiTag>,
+	mapApiTagToExportPath: Map<ApiTag, ExportData>,
+	logger: CommandLogger,
+): Map<ApiTag, ExportData> {
+	const result = new Map<ApiTag, ExportData>();
+	for (const [sourcePath, apiTag] of mapApiTagToSourcePath) {
+		// Check if apiTag exists in mapApiTagToOutputPath
+		const exportPath = mapApiTagToExportPath.get(apiTag);
+		if (exportPath) {
+			// Modify the exportPath
+			const modifiedExportPath = exportPath.relPath
+				.replace(/(lib|dist)/g, "src")
+				.replace(/\.js$|\.d\.ts$/, ".ts");
+
+			if (result.has(apiTag)) {
+				logger?.warning(`${modifiedExportPath} found in exports multiple times.`);
+			} else if (sourcePath === modifiedExportPath) {
+				result.set(apiTag, { ...exportPath, relPath: modifiedExportPath });
+			}
+		}
+	}
+
+	return result;
+}
+
+/**
+ * Read package "exports" to determine which "default"/ "types" paths to return along with `ApiTag`.
+ *
+ * @param packageJson - json content of package.json
+ * @param emitDeclarationOnly - If true, "types" exports are considered.
+ * @param logger - optional Logger
+ * @returns Map with API tags or levels with export path data
+ */
+function mapExportPathToApiTag(
+	packageJson: PackageJson,
+	tsconfig: TsConfigJson,
```

Comment:
TSDocs don't agree with arguments. The docs are better. Boolean that controls a search for "types" is better.